packages/protocol/contracts/common/UsingPrecompiles.sol

```diff
@@ -132,12 +159,104 @@ contract UsingPrecompiles {
    */
   function checkProofOfPossession(address sender, bytes memory blsKey, bytes memory blsPop)
     public
+    view
     returns (bool)
   {
     bool success;
-    (success, ) = PROOF_OF_POSSESSION.call.value(0).gas(gasleft())(
-      abi.encodePacked(sender, blsKey, blsPop)
-    );
+    (success, ) = PROOF_OF_POSSESSION.staticcall(abi.encodePacked(sender, blsKey, blsPop));
     return success;
   }
+
+  /**
+   * @notice Parses block number out of header.
+   * @param header RLP encoded header
+   * @return Block number.
+   */
+  function getBlockNumberFromHeader(bytes memory header) public view returns (uint256) {
+    bytes memory out;
+    bool success;
+    (success, out) = BLOCK_NUMBER_FROM_HEADER.staticcall(abi.encodePacked(header));
+    return getUint256FromBytes(out, 0);
+  }
+
+  /**
+   * @notice Computes hash of header.
+   * @param header RLP encoded header
+   * @return Header hash.
+   */
+  function hashHeader(bytes memory header) public view returns (bytes32) {
+    bytes memory out;
+    bool success;
+    (success, out) = HASH_HEADER.staticcall(abi.encodePacked(header));
+    return getBytes32FromBytes(out, 0);
+  }
+
+  /**
+   * @notice Gets the parent seal bitmap from the header at the given block number.
+   * @param blockNumber Block number to retrieve. Must be within 4 epochs of the current number.
+   * @return Bitmap parent seal with set bits at indices correspoinding to signing validators.
+   */
+  function getParentSealBitmap(uint256 blockNumber) public view returns (bytes32) {
+    bytes memory out;
+    bool success;
+    (success, out) = GET_PARENT_SEAL_BITMAP.staticcall(abi.encodePacked(blockNumber));
+    return getBytes32FromBytes(out, 0);
+  }
+
+  /**
+   * @notice Verifies the given header and returns the seal bitmap is successful.
+   * @param header RLP encoded header
+   * @return Bitmap parent seal with set bits at indices correspoinding to signing validators.
+   */
+  function getVerifiedSealBitmapFromHeader(bytes memory header) public view returns (bytes32) {
+    bytes memory out;
+    bool success;
+    (success, out) = GET_VERIFIED_SEAL_BITMAP.staticcall(abi.encodePacked(header));
+    return getBytes32FromBytes(out, 0);
+  }
+
+  /**
+   * @notice Converts bytes to address
+   * @param bs byte[] data
+   * @param start offset into byte data to convert
+   * @return address data
+   */
+  function getAddressFromBytes(bytes memory bs, uint256 start) internal pure returns (address) {
+    require(bs.length >= start + 32, "slicing out of range");
```

Comment:
Should we reuse the code in these three functions?

Reply:
Updated to reuse. Now with only 2 usages of "assembly" left, this PR addresses #1666.